node/lib/util/push.js

```diff
@@ -84,23 +84,75 @@ exports.push = co.wrap(function *(repo, remoteName, source, target, force) {
     let remoteUrl = yield GitUtil.getUrlFromRemoteName(repo, remoteName);
 
     // First, push the submodules.
+    const annotatedCommit = yield GitUtil.resolveCommitish(repo, source);
+    const sha = annotatedCommit.id();
+    const commit = yield repo.getCommit(sha);
+
+    const submoduleSet = new Set(yield SubmoduleUtil.listOpenSubmodules(repo));
+    for (const s of yield SubmoduleUtil.listAbsorbedSubmodules(repo)) {
+        submoduleSet.add(s);
+    }
+    const submodules = [...submoduleSet];
+    const pushMap = yield SubmoduleUtil.getSubmoduleShasForCommit(repo,
+                                                                  submodules,
+                                                                  commit);
+    const trackingBranches = new Set();
+    trackingBranches.add(`refs/remotes/${remoteName}/${target}`);
+    let tracking;
+    try {
+        const sourceBranch = yield repo.getReference(source);
+        tracking = yield GitUtil.getTrackingInfo(repo, sourceBranch);
+    } catch (e) {
+        // we have no local branch? maybe source is a sha.
+        tracking = null;
+    }
+
+    if (tracking !== null) {
+        if (tracking.remoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${remoteName}/${tracking.remoteName}`);
+        }
+        if (tracking.pushRemoteName !== null) {
+            trackingBranches.add(
+                `refs/remotes/${remoteName}/${tracking.pushRemoteName}`);
+        }
+    }
+
+    for (const branch of trackingBranches) {
+        const annotated = yield GitUtil.resolveCommitish(repo, branch);
```

Comment:
I think `resolveCommitish` is too heavy a tool for this; you have the exact ref name. I'd use `NodeGit.Reference.lookup`.